controller/src/main/java/io/pravega/controller/server/bucket/RetentionBucketService.java

```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.server.bucket;
+
+import io.pravega.client.stream.impl.StreamImpl;
+import io.pravega.common.concurrent.Futures;
+import io.pravega.common.hash.RandomFactory;
+import io.pravega.common.tracing.RequestTracker;
+import io.pravega.controller.store.stream.BucketStore;
+import io.pravega.controller.store.stream.OperationContext;
+import io.pravega.controller.store.stream.StreamMetadataStore;
+import io.pravega.controller.task.Stream.StreamMetadataTasks;
+import io.pravega.controller.util.Config;
+import io.pravega.controller.util.RetryHelper;
+import io.pravega.common.tracing.TagLogger;
+
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.Supplier;
+import org.slf4j.LoggerFactory;
+
+public class RetentionBucketService extends AbstractBucketService {
+    private static final TagLogger log = new TagLogger(LoggerFactory.getLogger(RetentionBucketService.class));
+
+    private final StreamMetadataTasks streamMetadataTasks;
+    private final StreamMetadataStore streamMetadataStore;
+    private final RequestTracker requestTracker;
+    private final Supplier<Long> requestIdGenerator = RandomFactory.create()::nextLong;
+
+    public RetentionBucketService(int bucketId, StreamMetadataStore streamMetadataStore, BucketStore bucketStore,
+                           StreamMetadataTasks streamMetadataTasks, ScheduledExecutorService executor,
+                           RequestTracker requestTracker) {
+        super(BucketStore.ServiceType.RetentionService, bucketId, bucketStore, executor);
+        this.streamMetadataTasks = streamMetadataTasks;
+        this.streamMetadataStore = streamMetadataStore;
+        this.requestTracker = requestTracker;
+    }
+    
+    CompletableFuture<Void> startWork(StreamImpl stream) {
+        // Randomly distribute retention work across RETENTION_FREQUENCY_IN_MINUTES spectrum by introducing a random initial
+        // delay. This will ensure that not all streams become eligible for processing of retention at around similar times.
+        long delay = Duration.ofMinutes(Config.MINIMUM_RETENTION_FREQUENCY_IN_MINUTES).toMillis();
+        long randomInitialDelay = ThreadLocalRandom.current().nextLong(delay);
+        return Futures.delayedFuture(() -> performRetention(stream), randomInitialDelay, executor)
+                      .thenCompose(x -> RetryHelper.loopWithDelay(this::isRunning, () -> performRetention(stream),
```

Comment:
Perhaps we should have a RetryHelper.loopWithRandomizedDelay() helper for this. That would also allow randomizing each iteration.